haystack/components/evaluators/exact_match_evaluator.py

```diff
@@ -0,0 +1,35 @@
+from itertools import product
+from typing import Any, Dict, List
+
+from haystack import default_from_dict, default_to_dict
+from haystack.core.component import component
+
+
+@component
+class ExactMatchEvaluator:
+    def to_dict(self) -> Dict[str, Any]:
+        return default_to_dict(self)
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "ExactMatchEvaluator":
+        return default_from_dict(cls, data)
+
+    @component.output_types(result=float)
+    def run(self, questions: List[str], ground_truths: List[List[str]], answers: List[List[str]]) -> Dict[str, float]:
```

Comment:
Can we expect this component to be used to compare retriever outputs? If so, it'd be a better idea to stick to `inputs, predictions, ground_truths` as the parameter names and then explain the concrete mapping between those names and the semantics depending on what this component will be evaluating.


Something like:
```python
def run(self, inputs: List[str], predictions: List[List[str]], ground_truths: List[List[str]]) -> Dict[str, float]:
	"""
	...
	
	:param inputs: Questions.
	:param predictions:
		When evaluating a retriever component, this would be the nested list of document contexts that were retrieved for each input question.
		When evaluating an E2E QA pipeline, this would be the nested list of generated or extracted answers.
	:param ground_truths:
		When evaluating a retriever component, this would be the nested list of expected document contexts for each input question.
		When evaluating an E2E QA pipeline, this would be the nested list of expected answers.
 	"""
```

Reply:
I would prefer questions over inputs too. ExactMatch is a metric that is for comparing answers not documents. So no retriever outputs, no.
And I am thinking, why not come up with a common terminology that can be applied to the integrations too? There we have `questions`, `contexts`, `responses`, `ground_truths`. https://github.com/deepset-ai/haystack-core-integrations/blob/0ccf07af3335df40009ba62fb77bc6059057a9ab/integrations/ragas/example/example.py#L44
What we can improve further is that we should distinguish ground truth responses from ground truth contexts and we should stick to `documents` instead of `contexts` and answers instead of `responses` in Haystack.

So what I'd suggest is that every metric can have a subset of `questions: List[str]`, `documents: List[List[Document]]`, `answers: List[List[Answer]]`,  `replies: List[List[str]]`, `ground_truth_documents: List[List[Document]]`, `ground_truth_answers: List[List[Answer]]`, `ground_truth_replies: List[List[str]]` as input. I added `replies` as this is what our generator components return.

Reply:
I don't want to discuss the naming in too much detail, that's easy too change later too. My main motivation is to make it intuitive for users. If they have a pipeline with a retriever and a generator, the outputs they get are called documents and and replies. That's what the evaluator should take as inputs then. And if the users has a reader component in the pipeline and the result is an answer of type ExtractAnswer, we should have an evaluator that takes that input.

Reply:
We could also keep `contexts: List[List[str]]` and then have a helper component that extracts the content from Documents.